x-pack/plugin/esql/qa/testFixtures/src/main/resources/hash.csv-spec

```diff
@@ -0,0 +1,14 @@
+hash
```

Comment:
I would be way more comfortable with more tests here. The one below is the bare minimum and could be extended with more examples. Look at this more as an educational suggestion, regarding the complex queries that can be built with ES|QL and `hash`.

- `FROM languages | eval x = ["md5", "foobar"] | EVAL hash(x, language_name)` this tests the multi-value handling of the functions
- `row x = "md5" | sort hash(x, x) | eval hash(x,x)` this is about `row` mostly
- `row x = ["md5","foobar"] | eval y = x | mv_expand x | eval hash(x,y)` multivalueness made single-value (thanks to `mv_expand`)
- `row x = "md",y = 5 | sort x, y | limit 1 | eval hash = concat("hash is ", hash(concat(x, y::string), y::string))` this is about nested calls to other functions; it's mostly about the type returned
- `from languages | eval hash(null, "foobar") | limit 3 | sort language_code` computing `hash` in a `from` (values from ES) but not actually using them. This is a scenario about constants in a non-constants (ES indices) setup
- `from *sample_data* metadata _index | eval h = hash("md5", @timestamp::long::string) | stats c = count(*) by h` this is about union types, meaning `@timestamp::long` which takes the `@timestamp` field from all the indices in the index pattern and converts all those different data types (date_nanos, date, long) to a common type `long`.

Trying to come up with interesting queries I have also found this one

```
from *sample_data* metadata _index | eval h = hash("md5", message) | stats c = count(*) by h | sort c desc nulls first
```

which _sometimes_ fails with the following error and stacktrace. Note though it fails sometimes (Windows OS here)

```
        "type": "array_index_out_of_bounds_exception",
        "reason": "arraycopy: length -20 is negative",
        "stack_trace": "java.lang.ArrayIndexOutOfBoundsException: arraycopy: length -20 is negative
```

and because it seems to fail because of the hashing algorithm I am mentioning here and you decide if this is about the `hash` function or about the algorithm itself.

```
status: 500 java.lang.ArrayIndexOutOfBoundsException: arraycopy: length -20 is negative
        at java.base/java.lang.System.arraycopy(Native Method)
        at java.base/sun.security.provider.DigestBase.engineUpdate(DigestBase.java:117)
        at java.base/sun.security.provider.MD5.implDigest(MD5.java:102)
        at java.base/sun.security.provider.DigestBase.engineDigest(DigestBase.java:205)
        at java.base/sun.security.provider.DigestBase.engineDigest(DigestBase.java:185)
        at java.base/java.security.MessageDigest$Delegate.engineDigest(MessageDigest.java:675)
        at java.base/java.security.MessageDigest.digest(MessageDigest.java:387)
        at org.elasticsearch.xpack.esql.expression.function.scalar.hash.Hash.hash(Hash.java:113)
        at org.elasticsearch.xpack.esql.expression.function.scalar.hash.Hash.processConstant(Hash.java:107)
        at org.elasticsearch.xpack.esql.expression.function.scalar.hash.HashConstantEvaluator.eval(HashConstantEvaluator.java:86)
        at org.elasticsearch.xpack.esql.expression.function.scalar.hash.HashConstantEvaluator.eval(HashConstantEvaluator.java:57)
        at org.elasticsearch.compute.operator.EvalOperator.process(EvalOperator.java:45)
        at org.elasticsearch.compute.operator.AbstractPageMappingOperator.getOutput(AbstractPageMappingOperator.java:76)
        at org.elasticsearch.compute.operator.Driver.runSingleLoopIteration(Driver.java:258)
        at org.elasticsearch.compute.operator.Driver.run(Driver.java:189)
        at org.elasticsearch.compute.operator.Driver$1.doRun(Driver.java:378)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:34)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:1023)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1575)
        Suppressed: java.lang.ArrayIndexOutOfBoundsException: arraycopy: length -20 is negative
```

Java version
```
java version "21.0.3" 2024-04-16 LTS
Java(TM) SE Runtime Environment (build 21.0.3+7-LTS-152)
Java HotSpot(TM) 64-Bit Server VM (build 21.0.3+7-LTS-152, mixed mode, sharing)
```

Reply:
Besides the ones mentioned by Andrei above, I'd add hashing of the empty string (in `HashTests`) and maybe providing the hash function directly in ROW (very optional, like `ROW h = HASH(...), ...`).

> I was assuming this is already covered by generated test cases from AbstractScalarFunctionTestCase.

These tests test the function in isolation. 

> what are the usual scenarios that we include into the csv-specs

It depends, for scalar functions I guess it may be less exhaustive, but I'd add providing input from fields (from an index, either directly or transformed), expressions (like `FROM employee | WHERE emp_no == 10005 | EVAL h = HASH(concat("MD", (emp_no%10)::STRING), first_name)`), MVs, in sorting expressions, filters, using in STATs. Some can be repetitive, but that's OK.

Reply:
@idegtiarenko 

> Re exception, by any chance do you have something that can reproduce it? (even if it is occasional)
> I have tried hundreds iterations with above sample and did not recorder failures: https://gradle-enterprise.elastic.co/s/y4aqo7okkvyvm/tests/overview

I do not. For me it reproduced every few query runs and after your latest updates to this PR (between my first review and this one now) it doesn't repro anymore and it seems to return correct results. I suggest adding that query to the csv-spec file along with the other queries from me and Bogdan (if not already added) and it should be good to go.